File: python/src/iap/labels/labels.py
"""Event-time forward labels at pinned horizons (conventions §7, spec §13).

Given a per-instrument event-time mid series (one sample per BOOK REFRESH:
``(exchange_ts, mid, half_spread, tradable)`` in price units) and a sorted
list of anchor timestamps (the feature emission times), this module computes,
for every anchor t and every horizon h in {10ms, 50ms, 100ms, 500ms, 1s, 5s,
10s, 30s, 1m, 5m, 15m}:

- ``mid``  : mid-to-mid forward return   m(t+h)/m(t) - 1
- ``cost`` : cost-adjusted forward return, crossing the half-spread at both
             ends (buy at the ask now, sell at the bid later):
             ((m(t+h) - hs(t+h)) - (m(t) + hs(t))) / m(t)

where m(t) / hs(t) is the mid / half-spread prevailing at t — the latest
series sample with ts <= t.

**Tradability (pinned, API_FEATURES §6).**  A series sample is ``tradable``
when, at that refresh, the merged book was two-sided (``book_ok``), NO venue
of the instrument was stale, and no venue reported HALT or AUCTION.  A label
is only a tradable forward return if the market was continuously observable
and open across the whole horizon, so a label at (t, h) is VALID only when
ALL of:

1. the stream was observed through t+h (``last_event_ts >= t + h``) —
   horizons running past the end of the session are invalid, never
   extrapolated (bit ``NOT_OBSERVED``);
2. an anchor sample exists with ``ts <= t``, its mid is > 0 (bit
   ``NO_ANCHOR``) and it is ``tradable`` (bit ``ANCHOR_NOT_TRADABLE``);
3. a forward sample exists with ``ts <= t + h`` (bit ``NO_FORWARD``), it is
   ``tradable`` and it is FRESH: ``t + h - ts_forward <= max_age_ns``
   (bit ``FORWARD_STALE``) — a mid frozen since the last quote of the day is
   not a price you could have traded at;
4. EVERY sample in ``(t, t + h]`` is tradable (bit ``BLACKOUT``) — a halt,
   a re-opening auction or a stale-venue gap anywhere inside the horizon
   invalidates the label, so a pre-halt signal is never credited with the
   reopen jump.

``max_age_ns`` is pinned as ``max(LABEL_MAX_AGE_FLOOR_NS, 2 x median
inter-sample gap of that instrument's series)`` (:func:`max_sample_age`):
a fixed floor of 5 s for dense equity books, scaled up for sparse FX
streams where a 15 s gap between LP quotes is normal, not a data outage.

Per-anchor reasons are returned as a bitmask (:class:`LabelReason`) so a
research report can say WHY a horizon has few usable rows.

No-lookahead alignment (pinned):

- the anchor state uses only events with exchange_ts <= t (the present);
- the forward state at t+h is the prevailing state after all events with
  exchange_ts <= t+h — i.e. it is determined exclusively by the anchor state
  plus events *strictly after* the anchor (mid changes between t and t+h).

Implementation: one two-pointer sweep per horizon plus a prefix count of
non-tradable samples — anchors are non-decreasing in time, so each of the 11
pointers only moves forward: O(len(anchors) * H + len(series)) total, no
per-anchor binary search.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Sequence

_NS_MS = 1_000_000
_NS_S = 1_000_000_000

#: Pinned label horizons (order is normative for serialized outputs).
HORIZONS_NS: Dict[str, int] = {
    "10ms": 10 * _NS_MS,
    "50ms": 50 * _NS_MS,
    "100ms": 100 * _NS_MS,
    "500ms": 500 * _NS_MS,
    "1s": 1 * _NS_S,
    "5s": 5 * _NS_S,
    "10s": 10 * _NS_S,
    "30s": 30 * _NS_S,
    "1m": 60 * _NS_S,
    "5m": 300 * _NS_S,
    "15m": 900 * _NS_S,
}
HORIZON_ORDER = tuple(HORIZONS_NS)

#: Freshness floor for the prevailing-mid age (pinned, see module docstring).
LABEL_MAX_AGE_FLOOR_NS = 5 * _NS_S
#: Multiplier applied to the median inter-sample gap.
LABEL_MAX_AGE_GAP_MULT = 2


class LabelReason:
    """Bit flags recorded per (anchor, horizon) when a label is invalid."""

    OK = 0
    NOT_OBSERVED = 1 << 0        # stream ended before t + h
    NO_ANCHOR = 1 << 1           # no prevailing sample at t, or mid <= 0
    ANCHOR_NOT_TRADABLE = 1 << 2  # book not two-sided / stale / halted at t
    NO_FORWARD = 1 << 3          # no prevailing sample at t + h
    FORWARD_STALE = 1 << 4       # prevailing mid at t + h older than max_age
    BLACKOUT = 1 << 5            # a non-tradable sample inside (t, t + h]

    NAMES = (
        (NOT_OBSERVED, "not_observed"),
        (NO_ANCHOR, "no_anchor"),
        (ANCHOR_NOT_TRADABLE, "anchor_not_tradable"),
        (NO_FORWARD, "no_forward"),
        (FORWARD_STALE, "forward_stale"),
        (BLACKOUT, "blackout"),
    )

    @classmethod
    def describe(cls, mask: int) -> List[str]:
        """Reason names set in ``mask`` (pinned order)."""
        return [name for bit, name in cls.NAMES if mask & bit]


@dataclass
class MidSeries:
    """Per-instrument event-time mid series (one sample per book refresh).

    ``tradable[i]`` is False for a refresh whose merged book was one-sided,
    whose instrument had a stale venue, or whose venues reported HALT /
    AUCTION.  Non-tradable samples carry ``mid = nan``.
    """

    ts: List[int] = field(default_factory=list)
    mid: List[float] = field(default_factory=list)
    half_spread: List[float] = field(default_factory=list)
    tradable: List[bool] = field(default_factory=list)

    def append(self, ts: int, mid: float, half_spread: float,
               tradable: bool = True) -> None:
        if self.ts and ts < self.ts[-1]:
            raise ValueError("MidSeries timestamps must be non-decreasing")
        self.ts.append(ts)
        self.mid.append(mid)
        self.half_spread.append(half_spread)
        self.tradable.append(bool(tradable))

    def __len__(self) -> int:
        return len(self.ts)

    def median_gap_ns(self) -> int:
        """Median gap between DISTINCT sample timestamps (0 with < 2).

        Several venue refreshes routinely share one ``exchange_ts`` (a
        consolidated book refreshes once per venue message); counting those
        zero gaps would collapse the median to 0 and make the freshness
        bound far too tight for a sparse FX stream.  The quote *cadence* is
        the gap between distinct instants.
        """
        seen: List[int] = []
        for t in self.ts:
            if not seen or t != seen[-1]:
                seen.append(t)
        if len(seen) < 2:
            return 0
        gaps = sorted(seen[i + 1] - seen[i] for i in range(len(seen) - 1))
        m = len(gaps)
        return int(gaps[m // 2] if m % 2 else (gaps[m // 2 - 1] + gaps[m // 2]) // 2)


def max_sample_age(series: MidSeries) -> int:
    """Pinned freshness bound for the prevailing mid of an instrument."""
    return max(LABEL_MAX_AGE_FLOOR_NS,
               LABEL_MAX_AGE_GAP_MULT * series.median_gap_ns())


@dataclass
class LabelResult:
    """Labels for one horizon across all anchors (parallel arrays)."""

    horizon: str
    mid: List[float]  # NaN where invalid
    cost: List[float]  # NaN where invalid
    valid: List[bool]
    reason: List[int] = field(default_factory=list)  # LabelReason bitmask


def compute_labels(
    anchors_ts: Sequence[int],
    series: MidSeries,
    last_event_ts: int,
    horizons: Sequence[str] = HORIZON_ORDER,
    max_age_ns: Optional[int] = None,
) -> Dict[str, LabelResult]:
    """Two-pointer forward-label sweep (see module docstring for semantics).

    ``anchors_ts`` must be non-decreasing.  ``last_event_ts`` is the
    timestamp of the last event observed for the instrument's stream.
    ``max_age_ns`` defaults to :func:`max_sample_age` of the series.
    """
    n = len(anchors_ts)
    for i in range(1, n):
        if anchors_ts[i] < anchors_ts[i - 1]:
            raise ValueError("anchors_ts must be non-decreasing")
    for h in horizons:
        if h not in HORIZONS_NS:
            raise ValueError(f"unknown horizon {h!r}")
    if max_age_ns is None:
        max_age_ns = max_sample_age(series)
    if max_age_ns <= 0:
        raise ValueError("max_age_ns must be positive")

    ts = series.ts
    mids = series.mid
    hss = series.half_spread
    tradable = series.tradable
    m = len(ts)
    nan = float("nan")

    # prefix count of NON-tradable samples: bad[i] = count over ts[:i]
    bad = [0] * (m + 1)
    for i in range(m):
        bad[i + 1] = bad[i] + (0 if tradable[i] else 1)

    # base pointer: latest series index with ts <= anchor
    out: Dict[str, LabelResult] = {}
    base_idx = [-1] * n
    j = -1
    for i in range(n):
        t = anchors_ts[i]
        while j + 1 < m and ts[j + 1] <= t:
            j += 1
        base_idx[i] = j

    for h in horizons:
        h_ns = HORIZONS_NS[h]
        lab_mid = [nan] * n
        lab_cost = [nan] * n
        lab_valid = [False] * n
        lab_reason = [0] * n
        k = -1  # latest series index with ts <= anchor + h
        for i in range(n):
            t = anchors_ts[i]
            target = t + h_ns
            while k + 1 < m and ts[k + 1] <= target:
                k += 1
            b = base_idx[i]
            reason = 0
            if last_event_ts < target:
                reason |= LabelReason.NOT_OBSERVED
            if b < 0:
                reason |= LabelReason.NO_ANCHOR
            elif not tradable[b]:
                reason |= LabelReason.ANCHOR_NOT_TRADABLE
            elif not (mids[b] > 0.0):
                reason |= LabelReason.NO_ANCHOR
            if k < 0:
                reason |= LabelReason.NO_FORWARD
            else:
                if not tradable[k] or not (mids[k] > 0.0):
                    reason |= LabelReason.BLACKOUT
                if target - ts[k] > max_age_ns:
                    reason |= LabelReason.FORWARD_STALE
            if b >= 0 and k >= b and bad[k + 1] - bad[b + 1] > 0:
                reason |= LabelReason.BLACKOUT
            lab_reason[i] = reason
            if reason:
                continue
            m0, hs0 = mids[b], hss[b]
            m1, hs1 = mids[k], hss[k]
            lab_mid[i] = m1 / m0 - 1.0
            lab_cost[i] = ((m1 - hs1) - (m0 + hs0)) / m0
            lab_valid[i] = True
        out[h] = LabelResult(horizon=h, mid=lab_mid, cost=lab_cost,
                             valid=lab_valid, reason=lab_reason)
    return out

File: rust/alpha/src/labels.rs
//! Event-time forward mid labels + Pearson IC (validation support for the
//! alpha golden group; semantics pinned by `iap/labels/labels.py` and
//! `iap/validation/metrics.py`).
//!
//! `mid_label(t, h) = m(t+h)/m(t) - 1` where `m(x)` is the mid prevailing
//! at-or-before `x` — ONE sample per book REFRESH, each carrying a
//! `tradable` flag (two-sided merged book, no stale venue, no HALT/AUCTION).
//!
//! A label is valid only when (API_FEATURES.md §6, pinned):
//! 1. the stream was observed through `t + h`;
//! 2. the anchor sample exists, is tradable and has a positive mid;
//! 3. the forward sample exists, is tradable and is FRESH
//!    (`t + h - ts_forward <= max_age_ns`);
//! 4. every sample in `(t, t + h]` is tradable (no halt / auction / stale
//!    gap inside the horizon).
//!
//! `max_age_ns` is `max(5 s, 2 x median gap between DISTINCT sample
//! timestamps)` — see [`max_sample_age`].

/// Freshness floor for the prevailing-mid age (pinned).
pub const LABEL_MAX_AGE_FLOOR_NS: i64 = 5_000_000_000;

/// Per-instrument event-time mid series (one sample per book REFRESH).
#[derive(Debug, Default, Clone)]
pub struct MidSeries {
    ts: Vec<i64>,
    mid: Vec<f64>,
    tradable: Vec<bool>,
    /// prefix count of NON-tradable samples (`bad[i]` covers `ts[..i]`)
    bad: Vec<u32>,
}

impl MidSeries {
    /// New empty series.
    pub fn new() -> MidSeries {
        MidSeries::default()
    }

    /// Append one TRADABLE sample (timestamps non-decreasing).
    pub fn append(&mut self, ts: i64, mid: f64) {
        self.append_sample(ts, mid, true);
    }

    /// Append one sample; `tradable == false` marks a blackout refresh
    /// (one-sided book, stale venue, HALT / AUCTION).
    pub fn append_sample(&mut self, ts: i64, mid: f64, tradable: bool) {
        debug_assert!(self.ts.last().map_or(true, |&last| ts >= last));
        let prev = self.bad.last().copied().unwrap_or(0);
        self.ts.push(ts);
        self.mid.push(mid);
        self.tradable.push(tradable);
        self.bad.push(prev + u32::from(!tradable));
    }

    /// Pinned freshness bound: `max(5 s, 2 x median gap between DISTINCT
    /// sample timestamps)` — a fixed floor for dense books, scaled up for a
    /// sparse FX stream where a 15 s quote gap is the normal cadence.
    pub fn max_sample_age(&self) -> i64 {
        let mut distinct: Vec<i64> = Vec::with_capacity(self.ts.len());
        for &t in &self.ts {
            if distinct.last() != Some(&t) {
                distinct.push(t);
            }
        }
        if distinct.len() < 2 {
            return LABEL_MAX_AGE_FLOOR_NS;
        }
        let mut gaps: Vec<i64> = distinct.windows(2).map(|w| w[1] - w[0]).collect();
        gaps.sort_unstable();
        let m = gaps.len();
        let median = if m % 2 == 1 {
            gaps[m / 2]
        } else {
            (gaps[m / 2 - 1] + gaps[m / 2]) / 2
        };
        LABEL_MAX_AGE_FLOOR_NS.max(2 * median)
    }

    /// Index of the latest sample at-or-before `t`.
    fn index_at_or_before(&self, t: i64) -> Option<usize> {
        let i = self.ts.partition_point(|&x| x <= t);
        if i > 0 {
            Some(i - 1)
        } else {
            None
        }
    }

    /// Number of samples.
    pub fn len(&self) -> usize {
        self.ts.len()
    }

    /// True when empty.
    pub fn is_empty(&self) -> bool {
        self.ts.is_empty()
    }

}

/// Forward mid labels for every anchor at one horizon; NaN where invalid
/// (see the module docstring for the pinned validity rules).
pub fn mid_labels(
    anchors_ts: &[i64],
    series: &MidSeries,
    last_event_ts: i64,
    horizon_ns: i64,
) -> Vec<f64> {
    let max_age = series.max_sample_age();
    mid_labels_with_age(anchors_ts, series, last_event_ts, horizon_ns, max_age)
}

/// [`mid_labels`] with an explicit freshness bound.
pub fn mid_labels_with_age(
    anchors_ts: &[i64],
    series: &MidSeries,
    last_event_ts: i64,
    horizon_ns: i64,
    max_age_ns: i64,
) -> Vec<f64> {
    anchors_ts
        .iter()
        .map(|&t| {
            let target = t + horizon_ns;
            if last_event_ts < target {
                return f64::NAN;
            }
            let (Some(b), Some(k)) = (
                series.index_at_or_before(t),
                series.index_at_or_before(target),
            ) else {
                return f64::NAN;
            };
            if !series.tradable[b] || !(series.mid[b] > 0.0) {
                return f64::NAN;
            }
            if !series.tradable[k] || !(series.mid[k] > 0.0) {
                return f64::NAN;
            }
            if target - series.ts[k] > max_age_ns {
                return f64::NAN;
            }
            if k >= b && series.bad[k] - series.bad[b] > 0 {
                return f64::NAN; // a blackout sample inside (t, t + h]
            }
            series.mid[k] / series.mid[b] - 1.0
        })
        .collect()
}

/// Pearson information coefficient over pairwise-finite entries; NaN with
/// fewer than `min_obs` pairs or a degenerate (std <= 1e-12) side. Mirrors
/// `iap.validation.metrics.ic` (default `min_obs = 32`).
pub fn pearson_ic(scores: &[f64], labels: &[f64], min_obs: usize) -> f64 {
    assert_eq!(scores.len(), labels.len(), "score/label length mismatch");
    let pairs: Vec<(f64, f64)> = scores
        .iter()
        .zip(labels.iter())
        .filter(|(x, y)| x.is_finite() && y.is_finite())
        .map(|(&x, &y)| (x, y))
        .collect();
    let n = pairs.len();
    if n < min_obs {
        return f64::NAN;
    }
    let nf = n as f64;
    let mx = pairs.iter().map(|p| p.0).sum::<f64>() / nf;
    let my = pairs.iter().map(|p| p.1).sum::<f64>() / nf;
    let mut sxx = 0.0;
    let mut syy = 0.0;
    let mut sxy = 0.0;
    for &(x, y) in &pairs {
        let (dx, dy) = (x - mx, y - my);
        sxx += dx * dx;
        syy += dy * dy;
        sxy += dx * dy;
    }
    // population std guard, as in the reference
    if (sxx / nf).sqrt() <= 1e-12 || (syy / nf).sqrt() <= 1e-12 {
        return f64::NAN;
    }
    sxy / (sxx.sqrt() * syy.sqrt())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn labels_use_at_or_before_and_session_end() {
        let mut s = MidSeries::new();
        s.append(0, 100.0);
        s.append(1_000, 101.0);
        s.append(3_000, 99.0);
        // anchor 500 with horizon 1000: m(500)=100 (at-or-before), m(1500)=101
        let lab = mid_labels(&[500], &s, 3_000, 1_000);
        assert!((lab[0] - (101.0 / 100.0 - 1.0)).abs() < 1e-15);
        // anchor 2500 with horizon 1000: target 3500 beyond last event -> NaN
        let lab = mid_labels(&[2_500], &s, 3_000, 1_000);
        assert!(lab[0].is_nan());
        // anchor before the first sample -> NaN
        let lab = mid_labels(&[-10], &s, 3_000, 5);
        assert!(lab[0].is_nan());
    }

    #[test]
    fn ic_matches_hand_computation_and_guards() {
        let x: Vec<f64> = (0..40).map(|i| i as f64).collect();
        let y: Vec<f64> = x.iter().map(|v| 2.0 * v + 1.0).collect();
        assert!((pearson_ic(&x, &y, 32) - 1.0).abs() < 1e-12);
        let yneg: Vec<f64> = x.iter().map(|v| -v).collect();
        assert!((pearson_ic(&x, &yneg, 32) + 1.0).abs() < 1e-12);
        // too few pairs
        assert!(pearson_ic(&x[..10], &y[..10], 32).is_nan());
        // degenerate side
        let flat = vec![1.0; 40];
        assert!(pearson_ic(&x, &flat, 32).is_nan());
        // NaNs dropped pairwise
        let mut y2 = y.clone();
        y2[3] = f64::NAN;
        assert!((pearson_ic(&x, &y2, 32) - 1.0).abs() < 1e-12);
    }
}
